sdks/go/pkg/beam/core/util/reflectx/call.go

```diff
@@ -20,44 +20,50 @@ import (
 	"reflect"
 	"sync"
 
+	"fmt"
+	"runtime/debug"
+
 	"github.com/apache/beam/sdks/go/pkg/beam/log"
 )
 
 //go:generate specialize --input=calls.tmpl
 
-// Caller is an untyped function call interface. This indirection allows
-// us to avoid reflection call overhead for certain types.
-type Caller interface {
+// Func represents a callable untyped function. This indirection allows
+// us to avoid reflection call overhead for certain types as well as
+// faster dynamic function implementations.
+type Func interface {
+	// Name returns the name of the function.
+	Name() string
 	// Type returns the type.
 	Type() reflect.Type
 	// Call invokes the implicit fn with arguments.
 	Call(args []interface{}) []interface{}
 }
 
 var (
-	callers   = make(map[string]func(interface{}) Caller)
-	callersMu sync.Mutex
+	funcs   = make(map[string]func(interface{}) Func)
+	funcsMu sync.Mutex
 )
 
-// RegisterCaller registers an custom caller factory for the given type, such as
-// "func(int)bool". If multiple caller factories are registered for the same type,
+// RegisterFunc registers an custom reflectFunc factory for the given type, such as
+// "func(int)bool". If multiple func factories are registered for the same type,
 // the last registration wins.
-func RegisterCaller(t reflect.Type, maker func(interface{}) Caller) {
-	callersMu.Lock()
-	defer callersMu.Unlock()
+func RegisterFunc(t reflect.Type, maker func(interface{}) Func) {
+	funcsMu.Lock()
+	defer funcsMu.Unlock()
 
 	key := t.String()
-	if _, exists := callers[key]; exists {
-		log.Warnf(context.Background(), "Caller for %v already registered. Overwriting.", key)
+	if _, exists := funcs[key]; exists {
+		log.Warnf(context.Background(), "Func for %v already registered. Overwriting.", key)
 	}
-	callers[key] = maker
+	funcs[key] = maker
 }
 
-// MakeCaller returns a caller for given function.
-func MakeCaller(fn interface{}) Caller {
-	callersMu.Lock()
-	maker, exists := callers[reflect.TypeOf(fn).String()]
-	callersMu.Unlock()
+// MakeFunc returns a reflectFunc for given function.
```

Comment:
<!--new_thread; commit:9c2390201ba3ecb080c84c7963396368ee720ae0; resolved:0-->
It's confusing to refer to an unexported type in godoc. Consider simply calling it Func instead.

Reply:
Done